src/llama.cpp

```diff
@@ -12670,7 +12687,17 @@ static void llama_set_inputs(llama_context & lctx, const llama_batch & batch) {
 
             GGML_ASSERT(ggml_backend_buffer_is_host(lctx.inp_KQ_mask->buffer));
 
-            float * data = (float *) lctx.inp_KQ_mask->data;
+            float * data     = (float *) lctx.inp_KQ_mask->data;
+            float * data_swa = nullptr;
+            const llama_pos n_keep_swa = hparams.n_sliding - batch.n_tokens;
+
+            if (lctx.model.arch == LLM_ARCH_GEMMA2) {
+                GGML_ASSERT(lctx.inp_KQ_mask_SWA);
+                GGML_ASSERT(hparams.n_sliding > 0);
+                data     = (float *) lctx.inp_KQ_mask->data;
+                data_swa = (float *) lctx.inp_KQ_mask_SWA->data;
+                // because layer masks are alternate for gemma 2, we only need to take first 2 layers
+            }
```

Comment:
This can be simplified a bit.
```suggestion
            if (lctx.inp_KQ_mask_SWA) {
                data_swa = (float *) lctx.inp_KQ_mask_SWA->data;
            }
```

Reply:
If I am not mistaken, mistral uses SWA every layer. So maybe this needs to be separated to allow having only `inp_KQ_mask_SWA`? Will the same implementation work?

Reply:
I've just looked at mistral reference implementation, they seem to use different mask for each layer. Link: https://github.com/mistralai/mistral-inference/blob/main/src/mistral_inference/cache.py

So I think my previous version (using `std::vector`) can handle that. Do you think I should revert the change?

It surprises me a bit, since mistral's quality doesn't seem to degrade even it's missing SWA (or it only breaks after 4096 tokens?)

Reply:
I have been looking at this code for a while and reviewing the mistral paper, and I think this is an implementation of the rolling buffer cache rather than sliding window attention. As far as I can tell, mistral has the same sliding window of 4096 tokens on each layer. Knowing that, it is possible to reduce the size of the KV cache to the sliding window size, but that requires some additional housekeeping so that eg. the rope still receives the absolute positions of the tokens, but the data is actually stored in the position `pos % sliding_window`. But maybe I am misunderstanding something, can you point me to the specific code?

Reply:
Yes, it should be possible. The thing I cannot figure out is how to avoid calling `llama_kv_cache_find_slot()` per-layer - seems it would be a big waste to do it like this, although it would generalize to support arbitrary KV cache layer sizes

Reply:
Maybe I misunderstood the code, but by looking at [generate.py](https://github.com/mistralai/mistral-inference/blob/main/src/mistral_inference/generate.py), `BufferCache` is initialized to `cache_window = max(seqlens) + max_tokens` which is not 4096 as I expected.

Then inside [cache.py](https://github.com/mistralai/mistral-inference/blob/main/src/mistral_inference/cache.py), what I understand is `cache_positions` should be caped for rolling (i.e. reset to 0 at some point) while `positions` keep increasing

```py
    # rope absolute positions
    positions: torch.Tensor
    # where tokens should go in the cache
    cache_positions: torch.Tensor
```

But that's not the case since `cache_positions` is only increased:

```py
cache_positions = positions + batch_idx * self.max_seq_len
```

Reply:
I am mostly going by the description in the [paper](https://arxiv.org/pdf/2310.06825), this code is basically gibberish to me.

Reply:
Yeah I assume the code is reference implementation so not very good quality. Having rolling buffer would be ideal in this case, but seems like too many changes. This is mostly to answer your question earlier: Will the same implementation work? Yes it works with different sliding window mask per layer, but will be waste of memory without rolling buffer.

Reply:
I looked deeper into the paper, seems like I missed something.

Looking at this figure:

![image](https://github.com/ggerganov/llama.cpp/assets/7702203/ec770a90-6843-4251-93ff-0587d188389e)

And the explanation:

![image](https://github.com/ggerganov/llama.cpp/assets/7702203/9b54f459-e6cf-43d2-84ee-f499ce462af2)

I'd assume that the mask for each layer is shifted by the size of `window - 1`, for example:
- layer 0: `0, 0, 0, 1, 1`
- layer 1: `0, 0, 1, 1, 0`
- layer 2: `0, 1, 1, 0, 0`
- ...

But then what I don't understand is the phrase "position i of the layer k, hi, attends to all hidden states from
the **previous** layer with positions between i − W and i". On the surface, it seems to explain how layer 1 knows about the tokens fall outside of its window (which is in layer 0), but then what's not clear to me is how one layer can attend to the previous one.

Also looking at the HF implementation code, seems like there is no such thing. They just add same attention mask for all layers: https://github.com/huggingface/transformers/blob/e65502951593a76844e872fee9c56b805598538a/src/transformers/models/mistral/modeling_mistral.py#L354

Reply:
> This can be simplified a bit.

Changed in https://github.com/ggerganov/llama.cpp/pull/8227/commits/ed5496fb32e9888abeaa7672aaba4d4251671457

I think for now we can keep the implementation this way, I'll need more time to figure out how mistral actually use SWA.

Reply:
> But then what I don't understand is the phrase "position i of the layer k, hi, attends to all hidden states from the **previous** layer with positions between i − W and i". On the surface, it seems to explain how layer 1 knows about the tokens fall outside of its window (which is in layer 0), but then what's not clear to me is how one layer can attend to the previous one.

I think it doesn't directly "attend" to the tokens from the previous one. It just receives information about those tokens through the output of previous layer. 

I am also trying to understand this concept from the past 3 days. I did not pay attention to this when Mistral v1 was released and I remember seeing that Mistral v2 removed SWA.